### DISCUSSION TITLE
Support for Vec in Query?

### DISCUSSION THREAD

**Discussion by Vesafary** [tokio-rs/axum#433] [Q&A] [answered]
Hi! 
I've been struggling with this for a while, and didn't get it to work. 
Looking through the docs and examples also didn't give me the answer.

What I'm looking for is support for one of the following:
```
<url>?id=1&id=2&id=3
<url>?id=[1, 2, 3]
```

I've tried several options like
```
struct IdQuery {
    ids: Vec<uuid::Uuid>,
}
async fn get_ids(filter: Query<IdQuery>)
```
```
struct IdQuery {
    id: uuid::Uuid,
}
async fn get_ids(filter: Query<Vec<IdQuery>>)
```
Of course I can implement it myself using the RawQuery extractor, but it seems like fairly basic usage so I was wondering whether I did something wrong.

Cheers!

**Comment by davidpdrsn** [comment]
Oh your right! axum uses serde_urlencoded for `Query` which it seems doesn't support sequences https://github.com/nox/serde_urlencoded/issues/85. I didn't realize this, thats quite a bummer.

I did a very quick test with `serde_qs` which is mentioned in the issue and that works:

```rust
struct Qs<T>(T);

#[axum::async_trait]
impl<T> FromRequest for Qs<T>
where
    T: serde::de::DeserializeOwned,
{
    type Rejection = Infallible;

    async fn from_request(req: &mut RequestParts) -> Result<Self, Self::Rejection> {
        // TODO: error handling
        let query = req.uri().query().unwrap();
        Ok(Self(serde_qs::from_str(query).unwrap()))
    }
}

async fn handler(Qs(ids): Qs<Input>) {
    dbg!(ids);
}

#[derive(serde::Deserialize, Debug)]
struct Input {
    ids: Vec<i32>,
}
```

Now I'm wondering if axum should just change to `serde_qs` instead of `serde_urlencoded` 🤔 

I'm gonna convert this into an issue.

**Comment by davidpdrsn** [comment, marked as answer]
See https://github.com/tokio-rs/axum/issues/434.

**Reply by Vesafary** [reply]
Thanks for the quick reply, I'll try this :)
